Make is_ascii_hexdigit branchless

GKFX opened the pull request:
Relevant issue #72895.

Use a bitwise or with `0x20` to make uppercase letters lowercase before comparing against the range `'a'..='f'`. This offers a significant speedup on my machine in a simple benchmark.

```rust
#![feature(test)]

extern crate test;

const RAND: &'static [u8; 500000] = include_bytes!("random.dat");

#[bench]
fn is_ascii_hex_current(bench: &mut test::Bencher) {
    bench.iter(|| {
        let mut total = 0;
        for byte in RAND.iter() {
            if byte.is_ascii_hexdigit() { total += 1; }
        }
        total
    });
}

#[bench]
fn is_ascii_hex_bitor(bench: &mut test::Bencher) {
    bench.iter(|| {
        let mut total = 0;
        for byte in RAND.iter() {
            if matches!(*byte, b'0'..=b'9') || matches!(*byte | 0x20, b'a'..=b'f') { total += 1; }
        }
        total
    });
}
```

```
test is_ascii_hex_bitor   ... bench:     222,629 ns/iter (+/- 1,426)
test is_ascii_hex_current ... bench:   1,531,543 ns/iter (+/- 25,991)
```

The generated code can also be seen to be branch-free: https://rust.godbolt.org/z/T57hG18hc.

(For reasons which aren't clear to me, `is_ascii_alphanumeric` gets optimized better despite looking very similar so I haven't made the corresponding change there.)

GKFX committed:
Make is_ascii_hexdigit branchless

Bitwise-or with 0x20 before checking if character in range a-z avoids
need to check if it is in range A-Z. This makes the generated code
shorter and faster.

rust-highfive commented:
r? @Mark-Simulacrum

(rust-highfive has picked a reviewer for you, use r? to override)

GKFX committed:
Fix typo in range

joshtriplett commented on library/core/src/char/methods.rs:
```diff
@@ -1510,7 +1510,8 @@ impl char {
     pub const fn is_ascii_hexdigit(&self) -> bool {
-        matches!(*self, '0'..='9' | 'A'..='F' | 'a'..='f')
+        // Bitwise or can avoid need for branches in compiled code.
+        matches!(*self, '0'..='9') || matches!(*self as u32 | 0x20, 0x61..=0x66)
```
To make this more maintainable, how about using `(b'a' as u32)..=(b'f' as u32)` instead?

GKFX replied:
Unfortunately that would be a syntax error. I can't figure out a nice-looking way to use character literals here; `RangeInclusive::contains` isn't const, `('a' as u32 <= *self as u32 | 0x20) && ('f' as u32 >= *self as u32 | 0x20)` is very long.

cuviper replied:
You can still use that range in a pattern if you define it as a `const` item first.

eggyal replied:
One could make it slightly more concise:

```rust
let lower = *self as u32 | 0x20;
matches!(*self, '0'..='9') || (lower >= b'a' as u32 && lower <= b'f')
```

As an aside, there's a (currently private) `const ASCII_CASE_MASK: u8` in `core::num` which might be preferable to a literal `0x20`.

GKFX replied:
> You can still use that range in a pattern if you define it as a `const` item first.

```
    const LOWER_ASCII: RangeInclusive<u32> = ('a' as u32)..=('f' as u32);
    matches!(*c, '0'..='9') || matches!(*c as u32 | 0x20, LOWER_ASCII)
```
[doesn't seem to compile](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d5c26475e124614b28e4bb4da5d40981)? I can get it to work with `A` and `F` const items though which I have done.

joshtriplett commented:
@rustbot author

scottmcm commented on library/core/src/char/methods.rs:
```diff
@@ -1510,7 +1510,8 @@ impl char {
     #[inline]
     pub const fn is_ascii_hexdigit(&self) -> bool {
-        matches!(*self, '0'..='9' | 'A'..='F' | 'a'..='f')
+        // Bitwise or can avoid need for branches in compiled code.
```
nit: this comment says "bitwise" but the next line uses *logical* or.  I don't know which is correct, but it seems like it could be improved to be more consistent.

GKFX replied:
The comment refers to the the bitwise or in `*self as u32 | 0x20`, which lets you check for uppercase and lowercase ASCII at once and therefore results in faster code. I can clarify the comment.

GKFX committed:
Clarify comment and use character literals in char::is_ascii_hexdigit

GKFX commented:
@rustbot ready

clarfonthey commented on library/core/src/char/methods.rs:
```diff
@@ -1510,7 +1510,10 @@ impl char {
     #[rustc_const_stable(feature = "const_ascii_ctype_on_intrinsics", since = "1.47.0")]
     #[inline]
     pub const fn is_ascii_hexdigit(&self) -> bool {
-        matches!(*self, '0'..='9' | 'A'..='F' | 'a'..='f')
+        // Bitwise or converts A-Z to a-z, avoiding need for branches in compiled code.
+        const A: u32 = 'a' as u32;
+        const F: u32 = 'f' as u32;
+        matches!(*self, '0'..='9') || matches!(*self as u32 | 0x20, A..=F)
```
Does this end up optimising the same if you refactor it to just call the `u8` method? Something like:

```rust
u8::try_from(*self).map_or(false, |c| c.is_ascii_hexdigit())
```

scottmcm replied:
No, because it has a branch for the "is it < 256?" part: <https://rust.godbolt.org/z/x13Yqx7Kx>

That said, something like that might actually perform really well, because it'd probably branch predict well.

For example, imagine something like this:
```rust
if *self <= 'f' {
    let c = u8::try_from(*this).unwrap();
    c.is_ascii_hexdigit()
} else {
    false
}
```

Yes there's a jump, but it probably predicts *great*.  Or at least as well as whatever jump you get from checking this condition in the first place...

clarfonthey replied:
Ah, I was hoping that LLVM would be smart enough to optimise that out (since all other paths also involve a <N<256 comparison) but I guess not.

scottmcm replied:
I filed an LLVM bug about this: https://github.com/llvm/llvm-project/issues/60683

Even in the simpler case of `is_ascii_digit` it doesn't do as well as it ought to.

clarfonthey commented on library/core/src/num/mod.rs:
```diff
@@ -688,7 +688,8 @@ impl u8 {
     #[inline]
     pub const fn is_ascii_hexdigit(&self) -> bool {
-        matches!(*self, b'0'..=b'9' | b'A'..=b'F' | b'a'..=b'f')
+        // Bitwise or converts A-Z to a-z, avoiding need for branches in compiled code.
```
Minor nit, but it feels like the lack of branches still isn't clear to the passive viewer since `'0'..='9'` and `'a'..='f'` are still two separate ranges with a gap in the middle.

Oh, and uh, that `||` is looking suspiciously not-branchless. Since it ends up being so in the resulting assembly, perhaps it could be reworded as `|`?

Reading the compiled output:
```asm
example::char_is_hex_2:
        mov     eax, dword ptr [rdi]
        lea     ecx, [rax - 48]
        cmp     ecx, 10
        setb    cl
        or      eax, 32
        add     eax, -97
        cmp     eax, 6
        setb    al
        or      al, cl
        ret
```

Which is effectively:
```rust
(*self - 48 < 10) | ((*self | 0x20) - 97 < 6)
```

It's less that there are "fewer" branches in this code, but more that going from three ranges to two triggers a threshold in LLVM's side that makes it decide that branches are no longer worth it, and it removes them.

So... tying this all together, maybe the real point is not that it's branchless by itself, but that there are fewer computations overall _and_ LLVM is likely to optimise it without branches as a result.

GKFX replied:
That seems reasonable, I can do one where all the optimization is done by hand to make it clear what the resulting assembly should be.

scottmcm replied:
Sorry for pushing churn on you here, but I'm actually going to ask that you aim for a middle ground where the code is still as obvious as possible while still producing branchless assembly.

Particularly, there's no need to replace the range changes with wrapping_sub manually, as LLVM will do that quite reliably: <https://rust.godbolt.org/z/rPaxbf1o7>

It's normal in the library for the code to *not* be in the form of the expected assembly.  For examples, `is_power_of_two` is phrased using `count_ones`, not the well-known bitwise tricks

https://github.com/rust-lang/rust/blob/1ca6777c014813e3bdb98d155562fc3d111d86dd/library/core/src/num/uint_macros.rs#L2132-L2134

as that leaves it up to LLVM to generate the best assembly sequence -- which *will* be the `(x != 0) && (x & (x-1)) == 0` on i386, but libcore doesn't need to know that.

So please go back to the version with the `const`s you had in a previous iteration.  And consider making a local for the lower-cased version of the character -- that would help localize the comment.  Maybe something like

```rust
        // Bitwise or converts A-Z to a-z, allowing checking for the letters with a single range.
        let lower = *self as u32 | num::ASCII_CASE_MASK;
        const A: u32 = 'a' as u32;
        const F: u32 = 'f' as u32;
        // Not using logical or because the branch isn't worth it
        self.is_ascii_digit() | matches!(*self as u32 | 0x20, A..=F)
```

You could also consider splitting the hex_letter part into a separate (non-`pub`) function so that `is_ascii_hexdigit` becomes just
```rust
        // Not using logical or because the branch isn't worth it
        self.is_ascii_digit() | self.is_ascii_hexletter()
```
to isolate the conversion-and-range stuff to its own thing.

@rustbot author

scottmcm replied:
Hmm, it's just something about the *or pattern* in `matches!`.  This is also branchless: <https://rust.godbolt.org/z/9xqT38rz9>
```rust
fn is_ascii_digit(c: u8) -> bool {
    matches!(c, b'0'..=b'9')
}

fn is_ascii_hexletter(c: u8) -> bool {
    matches!(c, b'A'..=b'F') | matches!(c, b'a'..=b'f')
}

pub fn is_ascii_hexdigit(c: u8) -> bool {
    is_ascii_digit(c) | is_ascii_hexletter(c)
}
```

(And yup, it works for `char` too: <https://rust.godbolt.org/z/WvfsjMG6a>.)

clarfonthey replied:
Loving the investigative work here. Perhaps this is an actual codegen issue, since `matches!(x, P) | matches!(x, Q)` should ideally generate code rather close to `matches!(x, P | Q)`. It feels like more of a codegen quirk than an LLVM optimizer quirk, but I could be wrong.

scottmcm replied:
I dug in some more, and it's actually more interesting than I expected!

It looks like with the or-pattern it's tripping a different technique in LLVM. <https://rust.godbolt.org/z/zdW1T43T3>

Starting with the nice obvious
```rust
pub fn is_ascii_hexletter(c: char) -> bool {
    matches!(c, 'A'..='F' | 'a'..='f')
}
```
SimplifyCfg and SROA get it down to the nice simple short-circuiting
```llvm
define noundef zeroext i1 @example::is_ascii_hexletter(i32 noundef %c) unnamed_addr {
start:
  %_4 = icmp ule i32 65, %c
  %_5 = icmp ule i32 %c, 70
  %or.cond = and i1 %_4, %_5
  br i1 %or.cond, label %bb5, label %bb2

bb2:                                              ; preds = %start
  %_2 = icmp ule i32 97, %c
  %_3 = icmp ule i32 %c, 102
  %or.cond1 = and i1 %_2, %_3
  br i1 %or.cond1, label %bb5, label %bb4

bb5:                                              ; preds = %bb2, %start
  br label %bb6

bb4:                                              ; preds = %bb2
  br label %bb6

bb6:                                              ; preds = %bb4, %bb5
  %.0 = phi i8 [ 1, %bb5 ], [ 0, %bb4 ]
  %0 = trunc i8 %.0 to i1
  ret i1 %0
}
```
and InstCombine does the "I know how to simplify range checks like that" rewrite:
```llvm
define noundef zeroext i1 @example::is_ascii_hexletter(i32 noundef %c) unnamed_addr {
start:
  %0 = add i32 %c, -65
  %1 = icmp ult i32 %0, 6
  br i1 %1, label %bb5, label %bb2

bb2:                                              ; preds = %start
  %2 = add i32 %c, -97
  %3 = icmp ult i32 %2, 6
  br i1 %3, label %bb5, label %bb4

bb5:                                              ; preds = %bb2, %start
  br label %bb6

bb4:                                              ; preds = %bb2
  br label %bb6

bb6:                                              ; preds = %bb4, %bb5
  %.0 = phi i1 [ true, %bb5 ], [ false, %bb4 ]
  ret i1 %.0
}
```
But then something really interesting happens, and `SimplifyCfg` says "wait, that looks like a lookup table!", giving
```llvm
define noundef zeroext i1 @example::is_ascii_hexletter(i32 noundef %c) unnamed_addr {
start:
  switch i32 %c, label %bb4 [
    i32 102, label %bb6
    i32 101, label %bb6
    i32 100, label %bb6
    i32 99, label %bb6
    i32 98, label %bb6
    i32 97, label %bb6
    i32 70, label %bb6
    i32 69, label %bb6
    i32 68, label %bb6
    i32 67, label %bb6
    i32 66, label %bb6
    i32 65, label %bb6
  ]

bb4:                                              ; preds = %start
  br label %bb6

bb6:                                              ; preds = %start, %start, %start, %start, %start, %start, %start, %start, %start, %start, %start, %start, %bb4
  %.0 = phi i1 [ false, %bb4 ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ]
  ret i1 %.0
}
```
Then later SimplifyCfg looks at that *again* and say "wait, that's a weird switch, how about I do that with a shift instead?" by encoding the table into an `i38` (because `'A'` through `'f'` [is 38 values](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f7d86be1f8b8c5d1b42392d54d3e7d96)):
```llvm
define noundef zeroext i1 @example::is_ascii_hexletter(i32 noundef %c) unnamed_addr {
start:
  %switch.tableidx = sub i32 %c, 65
  %0 = icmp ult i32 %switch.tableidx, 38
  %switch.cast = zext i32 %switch.tableidx to i38
  %switch.shiftamt = mul i38 %switch.cast, 1
  %switch.downshift = lshr i38 -4294967233, %switch.shiftamt
  %switch.masked = trunc i38 %switch.downshift to i1
  %.0 = select i1 %0, i1 %switch.masked, i1 false
  ret i1 %.0
}
```
which is *also* branchless.  Just not quite as good a way.

GKFX committed:
Fully optimize is_ascii_hexdigit by hand

This should compile to the same thing as the previous commit (at a suitable optimization level) but makes it very clear what is intended.

scottmcm reviewed (changes requested):
Sorry, I can see how this got lost in the long conversation, but based on my investigation in https://github.com/rust-lang/rust/pull/103024#discussion_r1002756673 this can be written without needing to manually do the `wrapping_sub` checks -- LLVM will happily do that itself, given the opportunity.

So as described in that comment, I'd still like to see it written in a more obviously-correct rust way that still optimizes well, something like <https://rust.godbolt.org/z/eoaq4c37e>.

@rustbot author

the8472 commented:
Since we're trying to please LLVM, should this have a codegen test?

Dylan-DPC commented:
@GKFX any updates on this?

JohnCSimon commented:
@GKFX 

Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you are going to continue please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive
